packages/foundry/src/vault/strategy/autocompouding/liquidityPair/protocols/sushi/MiniChefV2Compounder.sol

```diff
@@ -0,0 +1,128 @@
+// SPDX-License-Identifier: GPL-3.0
+// Docgen-SOLC: 0.8.15
+
+pragma solidity ^0.8.15;
+import { LiquidityPairBase, ERC20 } from "../../LiquidityPairBase.sol";
+import { IUniswapRouterV2 } from "../../../../../../interfaces/external/uni/IUniswapRouterV2.sol";
+import { IUniswapV2Pair } from "../../../../../../interfaces/external/uni/IUniswapV2Pair.sol";
+import { IMiniChefV2 } from "../../../../../../interfaces/external/sushi/IMiniChefV2.sol";
+import { IRewarder } from "../../../../../../interfaces/external/IRewarder.sol";
+
+contract LiquidityPairCompounder is LiquidityPairBase {
+  constructor(
+    address _native,
+    address _lpPair,
+    address _swapRouter,
+    address _vault,
+    address _strategist,
+    address[] memory _protocolAddresses,
+    uint256[] memory _protocolUints,
+    address[][] memory _rewardsToNativeRoutes,
+    address[] memory _nativeToLp0Route,
+    address[] memory _nativeToLp1Route
+  ) public {
+    native = _native;
+    lpPair = _lpPair;
+    swapRouter = _swapRouter;
+
+    vault = _vault;
+    strategist = _strategist;
+
+    protocolAddresses = _protocolAddresses;
+    protocolUints = _protocolUints;
+
+    rewardsToNativeRoutes = _rewardsToNativeRoutes;
+
+    _setUp(rewardsToNativeRoutes, nativeToLp0Route, nativeToLp1Route);
+  }
+
+  /*//////////////////////////////////////////////////////////////
+                          SETUP
+    //////////////////////////////////////////////////////////////*/
+
+  // Give allowances for protocol deposit and rewardToken swaps.
+  function _giveAllowances() internal override {
+    address chef = protocolAddresses[0];
+
+    ERC20(lpPair).approve(chef, type(uint256).max);
+    ERC20(native).approve(swapRouter, type(uint256).max);
+    ERC20(lpToken0).approve(swapRouter, type(uint256).max);
+    ERC20(lpToken1).approve(swapRouter, type(uint256).max);
+  }
+
+  /*//////////////////////////////////////////////////////////////
+                          HARVEST LOGIC
+    //////////////////////////////////////////////////////////////*/
+
+  // Claim rewards from underlying protocol.
+  function _claimRewards() internal override {
+    address chef = protocolAddresses[0];
+    uint256 pid = protocolUints[0];
+
+    IMiniChefV2(chef).harvest(pid, address(this));
+  }
+
+  // Swap all rewards to native token.
+  function _swapRewardsToNative() internal override {
+    uint256 len = rewardsToNativeRoutes.length;
+    for (uint256 i; i < len; ++i) {
+      address reward = rewardsToNativeRoutes[i][0];
+      address[] memory rewardRoute = rewardsToNativeRoutes[i];
+      uint256 rewardAmount = ERC20(reward).balanceOf(address(this));
+      if (rewardAmount > 0) {
+        _uniV2Swap(swapRouter, rewardRoute, rewardAmount);
```

Comment:
could we make this _uniV2Swap a generalized swap function? I think it would be easier in the long run to have a few small contracts for different interfaces that just execute swap and we can mix and match them in this strategy.

Reply:
Nevermind, I misunderstood the comment. Yes I've gone and abstracted this.